### PR TITLE
[pipelining] RNG seed management for PP init

### LINKED ISSUE #139304: [RFC][Pipelining] RNG state communication, avoid seed checkpoint
### Background

During model initialization, each PP stage initializes their own portion of the model.  This is preferrable over initializing the model all in one place and then transferring it to each stage, due to the high memory requirements of initializing an entire large model in one place.

Ignoring random states, each PP stage would start from the same RNG seed and each model chunk could have similar or even identical weights.  This would likely lead to numerical issues during training.  It also makes loss-curves of PP training less comparable with loss-curves of non-PP training, which lowers confidence in PP training and adds obstacles to debugging. 

Currently, torchtitan implements a workaround where the whole model is initialized on CPU offline in one process, saved to disk, and then loaded by each pipeline stage.  (Loading does not require loading the whole model into ram, only the portion used locally).  This results in consistent model initialization but at a runtime and UX cost of generating and using the seed checkpoint.

### Proposal
Instead, we can manage the RNG state automatically inside torch.pipelining, to avoid identical weights.

**Option 1. Use different seeds per stage**

We could simply re-seed the RNG on each PP stage, without worrying about actual initialization parity with non-PP training.

This option is potentially faster since it allows each stage to initialize in parallel.  It should also gaurantee 'safe' initializations, since each stage would have no chance of duplicate or similar init weights.  However, this initialization would not 'match' the initialization values for non-PP training, leading to less comparability between loss curves.

**Option 2. Sequentially initialize**

After stage 0 initializes, it could summarize its own RNG states and send them to stage 1, which in turn loads the states and initializes, then saves new states and sends them to stage 2, etc.  This results in serializing initialization, which would lead to an observable slowdown at startup time, but also matches perfectly (bitwise) with the initialization values that would be observed in a non-PP training.

This option requires more complexity:
- virtual-stages means more than one round of communication across PP ranks is needed
- V-schedules means sometimes, neighboring stages exist on the same rank and thus seed communication should be skipped for those cases

Note: seed save/restore can be implemented like this: https://gist.github.com/wconstab/6bdea055eff9a7904fdae595c6cdac6e

It might be worth implementing both options and experimenting with the tradeoffs.  I suspect most users would prefer option 1 if its significantly faster, and we can prove it leads to equally accurate training.

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140026
* __->__ #139445

Add new 'seeded_module_init' API to PipelineScheduleMulti
- note: skipped for PipelineScheduleSingle for now. Could add it there
  too, but i want to delete that class and unify

Example:

```
schedule.seeded_module_init(
   # required arg, can be any fn that accepts 'module' as arg;
   # can be ModuleClass.init_weights
   my_initializer_fn,

   # optional args
   mode="serial",
   seed=1234
)
```

Implement both 'serial' and 'parallel' options described in #139304

'parallel' mode sets a unique seed per rank, but does not worry about
parity with non-PP initialization.  It is fastest becuase it
parallelizes initialization and uses no communication.

'serial' mode does initialization one stage at a time, communicating the
RNG states from rank to rank as needed so that the resulting
initialization is equivalent to non-PP initialization where the whole
model is initialized on one rank.

The overall initialization is delegated to the 'schedule' since it knows which
stages it owns, which ranks they are on, and what order they go in.
But the per-stage initialization is handled by a user-passed function
that can do anything it wants to the given module owned by that stage.

### CODE REVIEW

File: torch/distributed/pipelining/schedules.py [hunk 1]
```diff
@@ -1041,6 +1041,77 @@ def __init__(
                 "Simply stop passing it, and everything should still work fine."
             )
 
+    def seeded_module_init(
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:1044]: #bikeshed_me
"seeded_module_init"?  meh.  
I first used 'initialize_stage' but i realized there is a preexisting '_initialize_stage' API right below that is confusing.

File: torch/distributed/pipelining/schedules.py [hunk 2]
```diff
@@ -1041,6 +1041,77 @@ def __init__(
...
+        self,
+        stage_initializer: Callable[[torch.nn.Module], None],
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:1046]: should 'stage_initializer' be an optional argument?  I could check stage.submod for the presence of an 'init_weights' method, and if present, use that as the default.  Otherwise, if 'stage_initializer' is not provided, I would raise an error.

File: torch/distributed/pipelining/schedules.py [hunk 3]
```diff
@@ -1041,6 +1041,77 @@ def __init__(
...
+                        "Pipeline using serial init: rank %d stage %d using manual seed %d and sending RNG state to next rank...",
+                        self.rank,
+                        stage.stage_index,
+                        initial_seed,
+                    )
+                elif not prev_stage_is_rank_local:
+                    # need to know the size/dtype of the state tensors so might as well get current state then overwrite
+                    state_cpu = torch.get_rng_state().to(device)
+                    state_cuda = torch.cuda.get_rng_state(device=device).to(device)
+                    torch.distributed.recv(
```
Comment (wconstab) [inline torch/distributed/pipelining/schedules.py:1092]: is it overkill to mess with the CPU rng state?  I included it since in theory, someone could use CPU side to calculate something and then copy that to GPU for their tensor init.

File: torch/distributed/pipelining/stage.py [hunk 4]
```diff
@@ -173,6 +173,16 @@ def __init__(
                 f"Pipeline group size {self.group_size} cannot be larger than number of stages {self.num_stages}"
             )
 
+        def stage_global_rank(peer_rank):
```
Comment (wconstab) [inline torch/distributed/pipelining/stage.py:176]: Note: just moved this function from a derived class up to the base class, so it can be used from everywhere.

**Comment by kwen2501** [comment]
Does "parallel" mode need help from `torch.pipelining`?

**Comment by wconstab** [comment]
> Does "parallel" mode need help from torch.pipelining?

not strictly.  It only needs to know 'pp_rank' to set different seeds per pp-rank, and then do init as usual.  However, once adding an API to pipelining, I think the only use case  for 'parallel' would be to get back some performance if serial was too slow, and having it be a flag-flip seems nicer than having to edit your code to do init differently.

**Comment by kwen2501** [comment]
I don't have problem landing the proposed as experimental API. The description in the RFC indicates that a pipeline schedule can be of some help for serialized init and hence why attaching a util here. (Although, it could be still debatable whether pipeline schedule -- a "run" thing -- should get caught in model init, but anyway). The reason for suggesting making it experimental is so that we can be off the hook of maintenance for now.

That said, just thinking out loud:
Re "parallel" mode:
Per @wconstab's comment above, users can probably do it easily / cleanly in their code. So this does not motivate the proposed API originally.

Re "serialized" mode:
The proposed API has an assumption that user provide a single init function for all the stage models. How easy / general is that (e.g. would that need some amount of if-else in the init function? Anything else?)

Re customer base:
If there are only a small set of users that need serialized init, what's wrong with recommending checkpoint based init to them? It seems similar to a training recovery step or an initial load step in modeling tuning. And would it be more performant than serialized init since all ranks can then load from the checkpoint concurrently? 

My point being, for the majority of users, should we not give them a feeling that they should/could change they are doing today? i.e. set different seeds, init the modules, hook it with pipeline schedule, run, etc.

**Comment by wconstab** [comment]
I think we should run experiments at larger scale to see if serialized init is noticeably slow. If not, I would be happy to recommend it as the default way, and just consider 'parallel' mode as an escape hatch.

> My point being, for the majority of users, should we not give them a feeling that they should/could change they are doing today? i.e. set different seeds, init the modules, hook it with pipeline schedule, run, etc.

Users can freely ignore this helper API and do their own seeding if they want.

> The proposed API has an assumption that user provide a single init function for all the stage models.

Agreed that this is a bit weak in general.  However, if users follow the practice of writing 'init_weights' per submodule then it is quite easy to have a generalized top-level initializer that delegates.  That is what we have in titan.

In the titan code, 
image

I was thinking maybe we can get rid of the 'for m in model_parts' loop here, define a helper 'parallelize_and_initialize(module)' that can be reused in PP and non PP branch, then just have the overall code look like this:

```
    def parallelize_and_initialize(m):
            # apply SPMD-style PT-D techniques
            models_parallelize_fns[model_name](m, world_mesh, parallel_dims, job_config)
            m.to_empty(device=init_device)
            m.init_weights(buffer_device=buffer_device)
            m.train()

    # apply parallelisms and initialization
    if parallel_dims.pp_enabled:
        # apply PT-D Pipeline Parallel
        pp_schedule, model_parts = models_pipelining_fns[model_name](
            model, pp_mesh, parallel_dims, job_config, device, model_config, loss_fn
        )

       schedule.seeded_module_init(parallelize_and_initialize)

    else:
        parallelize_and_initialize(model)

        model_parts = [model]
```

i'm not totally sure its good to bundle the parallelize and initialize in this way, but it might be ok?

Anyway.. I think there are 2 points
(a) should we offer serial init?
(b) if we are going to have serial init, what should the API be?

**Review by tianyu-l — COMMENTED** [review]
Looks quite nice!! Had some inline questions.

File: torch/distributed/pipelining/schedules.py [hunk 5]
```diff
@@ -1041,6 +1041,77 @@ def __init__(
...
+
+        for stage in self._stages:
+            next_stage_is_rank_local = stage.stage_index + 1 in local_stage_indices
+            prev_stage_is_rank_local = stage.stage_index - 1 in local_stage_indices
+            device = stage.device
+
+            # recv, set RNG seed
+            if mode == "serial":
+                if stage.is_first:
+                    torch.manual_seed(initial_seed)
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:1081]: I don't understand. 0 is a valid seed too.

If we had initial_seed=None, what should i pass to manual_seed? I can't turn the current RNG state (tensor) back into a seed (int) so I am forced to use an initial seed value here I think.

File: torch/distributed/pipelining/schedules.py [hunk 6]
```diff
@@ -1041,6 +1041,77 @@ def __init__(
...
+        transferred to the rank holding the next stage and set before performing next stage init.  Note: when two stages
+        are colocated on the same rank (See Note: V-schedule) there is no need to transfer the RNG state.
+
+        Currently, supports CPU and CUDA RNG states.  Uses device index specified by the 'stage'.  In the future, we may
+        change how 'device' is passed in.
+        """
+        assert mode in ("parallel", "serial"), f"unsupported mode {mode}"
+        if mode == "parallel":
+            seed = initial_seed + self.rank
+            torch.manual_seed(seed)
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:1066]: For q1; it's up to the user code, but if they don't do anything then every rank would start with the same seed. 

For q2; I don't remember how dtensor manages rng. We should check.

File: torch/distributed/pipelining/schedules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/pipelining/schedules.py:1081]: > If we had initial_seed=None, what should i pass to manual_seed?

I was thinking just don't call `manual_seed` if the seed is None. The point is, without PP user can call `init_weights` without providing a seed. I wonder if we can reproduce the same values with PP.

[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:1081]: i got your point, yea I can make that change.  For 'parallel' mode we'd still need to convert 'None' to 0 but for 'serial' mode we can just avoid setting anything on  stage0 for 'None' and then propagate from there.

[the remark this answers is not in the document]
Comment (tianyu-l) [reply torch/distributed/pipelining/schedules.py:1081]: Then maybe we should change the function name? How about `seed_aware_init`

My idea is: we should follow the flow of "keep single-device semantics" -> "decide functionality" -> "design API (with minimum interface)". In this case, without PP users don't need to explicitly provide a seed here to do init (in torchtitan there is already a [config](https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L370) to globally set the seed); with PP we can just mimic that behavior.

**Comment by XilunWu** [comment]
After going through this PR, I believe DTensor RNG won't have impact on this because DTensor RNG is only used on DTensor random ops (e.g. `torch.distributed.tensor.randn`) while PP RNG is directly calling `torch.manual_seed`. There's no interaction between them.

File: test/distributed/pipelining/model_registry.py [hunk 7]
```diff
@@ -87,6 +87,12 @@ def forward(self, x):
         x = self.net2(x)
         return x
 
+    def init_weights(self):
+        torch.nn.init.xavier_uniform_(self.net1.weight)
+        torch.nn.init.xavier_uniform_(self.net2.weight)
+        torch.nn.init.zeros_(self.net1.bias)
+        torch.nn.init.zeros_(self.net2.bias)
```
Comment (XilunWu) [inline test/distributed/pipelining/model_registry.py:94]: is this the place where model init happens?

Comment (wconstab) [reply test/distributed/pipelining/model_registry.py:94]: for the test case, yes.  But 'self' could be already an FSDP model, i just didn't test it that way

**Review by H-Huang — APPROVED** [review]

File: torch/distributed/pipelining/schedules.py [hunk 8]
```diff
@@ -1041,6 +1041,77 @@ def __init__(
...
+            stage_initializer(stage.submod)
+
+            # send new RNG seed
+            if mode == "serial" and not next_stage_is_rank_local and not stage.is_last:
+                state_cpu = torch.get_rng_state().to(device)
+                state_cuda = torch.cuda.get_rng_state(device=device).to(device)
+                torch.distributed.send(
+                    state_cpu, dst=stage.next_rank, group=stage.group
+                )
+                torch.distributed.send(
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/distributed/pipelining/schedules.py:1111]: I don't think so, unless I have a bug. They should run in a deterministic order that is the same on all ranks (init first then shape).

**Comment by wconstab** [comment]
Abandoning this as not widely useful.  It works fine for PP only (1D parallel) but will not work as intended when combined with SPMD/DTensor parallelism.  That's because DTensor/SPMD parallelism does not have a way to gaurantee same initialization behavior under different sharding configurations.
